feat: add dictionary APIs and corresponding tests

poppoerika opened the pull request:
Port over the following Dictionary APIs in the .NET incubating repo to the JavaScript SDK.
- `dictionaryFetch`
- `dictionarySetField`
- `dictionarySetFields`
- `dictionaryGetField`
- `dictionaryGetFields`
- `dictionaryRemoveField`
- `dictionaryRemoveFields`
- `dictionaryIncrement`

Also add tests for these APIs.

In addition to adding the Dictionary APIs, I'm removing the value accessors from the `set` operation and corresponding tests as well as validators for key, value, field, and item and associated tests.

Closes #143

schwern commented on src/messages/responses/cache-dictionary-fetch.ts:
```diff
@@ -0,0 +1,63 @@
+import {SdkError} from '../../errors/errors';
+import {applyMixins, ErrorBody} from '../../errors/error-utils';
+import {TextDecoder} from 'util';
+import {cache} from '@gomomento/generated-types';
```
```suggestion
import {cache_client} from '@gomomento/generated-types/dist/cacheclient';
```

This reduces `cache.cache_client._Foo` to `cache_client._Foo`

schwern commented on integration/simple-cache-client.test.ts:
```diff
@@ -193,32 +194,6 @@ describe('SimpleCacheClient.ts Integration Tests', () => {
       expect(getResponse.valueString()).toEqual(cacheValue);
     }
   });
-  it('should set and get string from cache and returned set value matches string cacheValue', async () => {
```
Note in the PR that we're also removing this, please.

schwern commented on src/simple-cache-client.ts:
```diff
@@ -162,6 +163,21 @@ export class SimpleCacheClient {
+   * Fetch the entire dictionary from the cache.
+   * @param {string} cacheName - Name of the cache to perform the lookup in.
+   * @param {string} dictionaryName - The dictionary to fetch.
+   * @returns Promise<DictionaryFetch.Response> - Promise containing the result of the fetch operation and the associated dictionary.
+   * @memberof SimpleCacheClient
```
I think [types have to be in brackets](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns).

```suggestion
   * @returns {Promise<DictionaryFetch.Response>} - Promise containing the result of the fetch operation and the associated dictionary.
   * @memberof {SimpleCacheClient}
```

poppoerika commented on integration/simple-cache-client.test.ts:
```diff
@@ -386,3 +375,1633 @@ describe('Integration Tests for operations on sets datastructure', () => {
+    successResponse = response as CacheDictionaryIncrement.Success;
+    expect(successResponse.valueNumber()).toEqual(100);
+
+    // Reset field
+    await momento.dictionarySetField(
+      INTEGRATION_TEST_CACHE_NAME,
+      dictionaryName,
+      field,
+      '0'
+    );
+    response = await momento.dictionaryIncrement(
+      INTEGRATION_TEST_CACHE_NAME,
+      dictionaryName,
+      field,
+      0
+    );
+    expect(response).toBeInstanceOf(CacheDictionaryIncrement.Success);
+    successResponse = response as CacheDictionaryIncrement.Success;
+    expect(successResponse.valueNumber()).toEqual(0);
+  });
```
I have been puzzled by this part of the test keeps failing...
`dictionaryIncrement` succeeds but the passed value to `CacheDictionaryIncrement.Success` is somehow `undefined`.
This seems to only happen when I try to increment 0 by 0.
When I tried to increment 0 by 1 and increment 1 by 0, both succeeded.
Somehow only JS cannot increment 0 by 0 (there was no issue for the same test in the .NET and PHP SDKs).

poppoerika replied:
PHP SDK test: https://github.com/momentohq/client-sdk-php/blob/main/tests/Cache/CacheClientTest.php#L1204
.NET SDK test: https://github.com/momentohq/client-sdk-dotnet-incubating/blob/main/tests/Integration/Momento.Sdk.Incubating.Tests/DictionaryTest.cs#L210

schwern replied:
@poppoerika Was this fixed? If not, can we make a FF issue?

schwern reviewed (commented):
If you run `npm run format` it will change the code to it's liking and avoid those check warnings.

poppoerika commented on src/utils/collection-ttl.ts:
```diff
@@ -12,38 +12,31 @@
  */
 export class CollectionTtl {
-  private readonly _ttlSeconds: number | null;
+  private readonly _ttlSeconds: number | undefined;
```
I kept it as you implemented so the above change is not applicable from my PR.

poppoerika commented on src/utils/collection-ttl.ts:
```diff
@@ -12,38 +12,31 @@
    * @param {boolean} [refreshTtl=true]
    */
-  constructor(ttlSeconds: number | null = null, refreshTtl = true) {
+  constructor(ttlSeconds: number | undefined = undefined, refreshTtl = true) {
```
same comment for here

virratanasangpunth commented on src/messages/responses/cache-dictionary-fetch.ts:
```diff
@@ -0,0 +1,74 @@
+      const keyValue = `${key}: ${value}, `;
+      stringRepresentation = stringRepresentation + keyValue;
+    });
+    return `${super.toString()}: valueDictionaryStringString: ${stringRepresentation.slice(
+      0,
+      -2
+    )}`;
```
What's up with `slice(0, -2)`? Is this to remove the extra `, ` at the end?

Will this work with an empty dictionaryy?

might be worth writing a unit test for this toString() (don't have to cover all the cases, just super basic ones).

also might be worth grabbing the string truncate logic from Schwern's code (if applicable, i haven't put too much thoughts into this): https://github.com/momentohq/client-sdk-javascript/blob/0064bdece94623730362abccdaa48b5ff8b1e0d6/src/messages/responses/cache-list-fetch.ts#L39

poppoerika replied:
Using `slice(0, -2) on an empty string results in an empty string.
Even if `valueDictionaryStringString` returns an empty map, `toString()` will return an empty string (can be observed [here](https://jsfiddle.net/13s9vdeb/2/))

schwern replied:
There's code in listFetch that might be useful.

virratanasangpunth commented on src/simple-cache-client.ts:
```diff
@@ -189,6 +195,110 @@ export class SimpleCacheClient {
+    );
+  }
+
+  public async dictionarySetFields(
```
I think we need javadoc at the top of all these public methods.

virratanasangpunth reviewed (commented):
LGTM.

A few comments.

schwern commented on src/utils/validators.ts:
```diff
@@ -30,11 +36,7 @@ export function ensureValidSetRequest(
-  if (!value) {
-    throw new InvalidArgumentError('value must not be empty');
-  }
-
+  ensureValidValue(value);
```
This code was wrong. It should be possible to set a key and value of `''`. Don't carry that forward to the dictionary requests. It's also not necessary to check for nil, the types on the SimpleCacheClient methods prevent that.

I think ensureValidSetRequest should be inlined. It's [called in one place](https://github.com/momentohq/client-sdk-javascript/blob/a2eb281c17ccc32d03aca412b432d7b547d730a7/src/internal/cache-client.ts#L122).

schwern reviewed (changes requested):
I haven't looked at the whole thing, but...

* The set changes should be split into its own PR. We'll want to fix ensureSetRequest as well.
* The tests are great, but there's a lot of copy and paste which makes them difficult to follow. I'm working on a way to reduce that.

virratanasangpunth commented on integration/simple-cache-client.test.ts:
```diff
@@ -2159,21 +2159,21 @@ describe('Integration tests for dictionary operations', () => {
 
     // Commenting out the below reset field this for now and investigate: https://github.com/momentohq/client-sdk-javascript/issues/172
     // Reset field
```
nit: remove this comment too?

poppoerika commented on src/simple-cache-client.ts:
```diff
@@ -189,6 +196,172 @@ export class SimpleCacheClient {
+  public async dictionaryIncrement(
+    cacheName: string,
+    dictionaryName: string,
+    field: string,
```
The .NET SDK's dictionary increment only supports for string field. 
https://github.com/momentohq/client-sdk-dotnet-incubating/blob/main/src/Momento.Sdk.Incubating/SimpleCacheClient.cs#L251

schwern replied:
I think that's a bug. Every other dictionary method takes both strings and bytes.

poppoerika replied:
Michael L. might have a background story to this.
@malandis - Do we have specific reasons why we only allow a string value for `field` in `dictionaryIncrement`?

cprice404 replied:
For the value, we only support incrementing a field if the fields contents are a valid UTF-8 String representation of an integer, so there are reasons to consider having slightly different handling for the value on this operation.

However, if I'm reading this right, this suggestion refers to the field *name*, not the *value*, correct?  I agree with Schwern that there's no reason to have special rules for the field *name* that we don't apply on other operations.

poppoerika commented on src/utils/validators.ts:
```diff
@@ -46,6 +48,18 @@ export function ensureValidKey(key: unknown) {
   }
 }
 
+export function ensureValidField(field: unknown) {
+  if (!field) {
+    throw new InvalidArgumentError('field must not be empty');
+  }
+}
+
+export function ensureValidValue(value: unknown) {
+  if (!value) {
+    throw new InvalidArgumentError('value must not be empty');
+  }
+}
```
It looks like we don't allow an empty string for key/value: https://github.com/momentohq/client-sdk-dotnet-incubating/blob/main/src/Momento.Sdk.Incubating/SimpleCacheClient.cs#L251
Do we want to update that too?

poppoerika replied:
Even if so, I'm going to update only for what I've added in this PR. 
I don't want to add too many things since this is already a big PR 🙇‍♀️

schwern replied:
Those only appear to check for null. https://github.com/momentohq/client-sdk-dotnet-incubating/blob/1b648f27a3d13a4a8cc12fdea4f6656e3071f840/tests/Integration/Momento.Sdk.Incubating.Tests/DictionaryTest.cs#L19

(I don't know why nulls can even be passed to that method)

poppoerika commented on integration/simple-cache-client.test.ts:
```diff
@@ -579,3 +567,1634 @@ describe('Integration Tests for operations on sets datastructure', () => {
+  });
+
+  it('should return InvalidArgument response for dictionarySetField with invalid cache/dictionary/field/value name', async () => {
+    const dictionarySetFieldResponse1 = await momento.dictionarySetField(
```
I noticed that I should reuse the same variable in the later tests 😄

cprice404 replied:
I disagree with this feedback.  it's fine for this PR but I would like to align on this before we consider switching any other tests over to use mutable variables.  Not a blocker, not requesting any changes for now.

schwern replied:
I think this indicates a larger code smell: the test is doing too much. This will be DRYed up by my later PR.

poppoerika replied:
🙇‍♀️

poppoerika commented:
As mentioned in the description, I removed the validators for key, value, field, and item since they can be an empty string and they cannot be passed as args. 
I also removed the corresponding tests.
Feedback from Michael S. https://github.com/momentohq/client-sdk-javascript/pull/164#discussion_r1072976564

cprice404 commented on integration/simple-cache-client.test.ts:
```diff
@@ -579,3 +518,1513 @@ describe('Integration Tests for operations on sets datastructure', () => {
+
+describe('Integration tests for dictionary operations', () => {
+  it('should return InvalidArgument response for dictionaryGetField with invalid cache and dictionary names', async () => {
+    let response = await momento.dictionaryGetField(
```
I realize that you got conflicting feedback on this on an earlier PR comment, and these are tests and not production code or public-facing examples so I don't need you to change them for this PR, but except in very rare cases I feel strongly that `let` is a code smell and should be avoided.  You do not need mutable variables here.

In the future if there is a question about changing an immutable variable to a mutable one please ping me so that we can discuss, I do not want to set a precedent for using unnecessary mutable variables and have that get perpetuated throughout the code base.  cc: @schwern .

I'd love to see these changed back but it does not need to block this PR or be handled urgently.

cprice404 reviewed (commented):
I think that we should fix the type for the dictionary increment field as per Schwern's comment but other than that I think I'm +1 on this and would like to get it merged soon, assuming no one else sees remaining blockers.
